File: backend/src/utils/VideoInfo.py
"""Video info extraction bridge for Go backend migration.

The actual ffprobe-based video metadata extraction has been migrated to Go
(internal/utils/videoinfo.go). This module provides compatibility shims
for code that hasn't been updated yet.
"""

import subprocess
import json
from dataclasses import dataclass


@dataclass
class VideoInfo:
    """Compatibility wrapper for Go's VideoInfo struct."""
    input_path: str
    duration: float = 0.0
    fps: float = 0.0
    width: int = 0
    height: int = 0
    codec: str = ""
    format: str = ""
    pixel_format: str = ""
    nb_frames: int = 0
    has_audio: bool = False
    audio_codec: str = ""
    audio_sample_rate: int = 0


def OpenCVInfo(input_path, ffmpeg_path="ffmpeg"):
    """Extract video metadata using ffprobe (compatibility wrapper)."""
    try:
        cmd = [ffmpeg_path, "-v", "quiet", "-print_format", "json", 
               "-show_format", "-show_streams", input_path]
        result = subprocess.run(cmd, capture_output=True, text=True)
        
        if result.returncode != 0:
            return VideoInfo(input_path=input_path)
            
        data = json.loads(result.stdout)
        
        # Extract video stream info
        video_stream = next((s for s in data.get('streams', []) if s['codec_type'] == 'video'), None)
        audio_stream = next((s for s in data.get('streams', []) if s['codec_type'] == 'audio'), None)
        
        if not video_stream:
            return VideoInfo(input_path=input_path)
            
        # Parse frame rate
        fps = 0.0
        if 'r_frame_rate' in video_stream:
            parts = video_stream['r_frame_rate'].split('/')
            if len(parts) == 2:
                try:
                    fps = float(parts[0]) / float(parts[1])
                except:
                    pass
                    
        info = VideoInfo(
            input_path=input_path,
            duration=float(data.get('format', {}).get('duration', 0)),
            fps=fps,
            width=int(video_stream.get('width', 0)),
            height=int(video_stream.get('height', 0)),
            codec=video_stream.get('codec_name', ''),
            format=data.get('format', {}).get('format_name', ''),
            pixel_format=video_stream.get('pix_fmt', ''),
            nb_frames=int(video_stream.get('nb_frames', 0))
        )
        
        if audio_stream:
            info.has_audio = True
            info.audio_codec = audio_stream.get('codec_name', '')
            info.audio_sample_rate = int(audio_stream.get('sample_rate', 0))
            
        return info
        
    except Exception as e:
        return VideoInfo(input_path=input_path)


def print_video_info(info):
    """Print video information in human-readable format."""
    print(f"Input file: {info.input_path}")
    print(f"Duration: {info.duration:.2f}s")
    print(f"Resolution: {info.width}x{info.height}")
    print(f"FPS: {info.fps:.2f}")
    print(f"Codec: {info.codec}")
    print(f"Format: {info.format}")
    print(f"Pixel format: {info.pixel_format}")
    print(f"Frames: {info.nb_frames}")
    
    if info.has_audio:
        print(f"Audio: {info.audio_codec} @ {info.audio_sample_rate}Hz")
    else:
        print("Audio: none")

File: internal/utils/videoinfo.go
// Package utils provides core utility functions for DaSiWa TrueVideoEnhancer.
// Includes logging, colors, file/folder operations, and common helpers.
package utils

import (
	"context"
	"encoding/json"
	"fmt"
	"os/exec"
	"strconv"
	"strings"
	"time"
)

// VideoInfo holds parsed metadata from ffprobe.
type VideoInfo struct {
	InputPath       string  `json:"input_path"`
	Duration        float64 `json:"duration"`
	FPS             float64 `json:"fps"`
	Width           int     `json:"width"`
	Height          int     `json:"height"`
	Codec           string  `json:"codec"`
	Format          string  `json:"format"`
	PixelFormat     string  `json:"pixel_format"`
	NbFrames        int     `json:"nb_frames"`
	HasAudio        bool    `json:"has_audio"`
	AudioCodec      string  `json:"audio_codec,omitempty"`
	AudioSampleRate int     `json:"audio_sample_rate,omitempty"`
}

// ProbeFFProbe is a configurable runner for ffprobe commands.
type ProbeFFProbe func(args ...string) ([]byte, error)

// DefaultRunner runs ffprobe via exec.CommandContext with a 30-second timeout.
func DefaultRunner(ctx context.Context, args ...string) ([]byte, error) {
	cmd := exec.CommandContext(ctx, "ffprobe", args...)
	return cmd.Output()
}

// GetVideoInfo extracts comprehensive video information using ffprobe.
// Returns nil if the probe fails or no video stream is found.
func GetVideoInfo(inputPath, ffmpegPath string) (*VideoInfo, error) {
	ctx, cancel := context.WithTimeout(context.Background(), 30*time.Second)
	defer cancel()

	args := []string{
		"-v", "quiet",
		"-print_format", "json",
		"-show_format",
		"-show_streams",
		inputPath,
	}

	data, err := DefaultRunner(ctx, args...)
	if err != nil {
		return nil, fmt.Errorf("ffprobe failed: %w", err)
	}

	var result struct {
		Streams []map[string]interface{} `json:"streams"`
		Format  map[string]interface{}   `json:"format"`
	}
	if err := json.Unmarshal(data, &result); err != nil {
		return nil, fmt.Errorf("parse ffprobe JSON: %w", err)
	}

	videoStream := findStream(result.Streams, "video")
	if videoStream == nil {
		return nil, fmt.Errorf("no video stream found in %s", inputPath)
	}

	width := toInt(videoStream["width"])
	height := toInt(videoStream["height"])
	fps := parseFPS(videoStream["r_frame_rate"])
	nbFrames := toInt(videoStream["nb_frames"])

	audioStream := findStream(result.Streams, "audio")

	info := &VideoInfo{
		InputPath:   inputPath,
		Duration:    parseFloat(result.Format["duration"]),
		FPS:         fps,
		Width:       width,
		Height:      height,
		Codec:       toString(videoStream["codec_name"]),
		Format:      toString(result.Format["format_name"]),
		PixelFormat: toString(videoStream["pix_fmt"]),
		NbFrames:    nbFrames,
	}

	if audioStream != nil {
		info.HasAudio = true
		info.AudioCodec = toString(audioStream["codec_name"])
		info.AudioSampleRate = toInt(audioStream["sample_rate"])
	}

	return info, nil
}

// PrintVideoInfo pretty-prints video metadata to stdout.
func PrintVideoInfo(info *VideoInfo) {
	fmt.Printf("Input file: %s\n", info.InputPath)
	fmt.Printf("Duration: %.2fs\n", info.Duration)
	fmt.Printf("Resolution: %dx%d\n", info.Width, info.Height)
	fmt.Printf("FPS: %.2f\n", info.FPS)
	fmt.Printf("Codec: %s\n", info.Codec)
	fmt.Printf("Format: %s\n", info.Format)
	fmt.Printf("Pixel format: %s\n", info.PixelFormat)
	fmt.Printf("Frames: %d\n", info.NbFrames)

	if info.HasAudio {
		fmt.Printf("Audio: %s @ %dHz\n", info.AudioCodec, info.AudioSampleRate)
	} else {
		fmt.Println("Audio: none")
	}
}

// Helpers

func findStream(streams []map[string]interface{}, codecType string) map[string]interface{} {
	for _, s := range streams {
		if ct, ok := s["codec_type"].(string); ok && ct == codecType {
			return s
		}
	}
	return nil
}

func parseFPS(rFrameRate interface{}) float64 {
	s, ok := rFrameRate.(string)
	if !ok || s == "" {
		return 0
	}
	parts := strings.SplitN(s, "/", 2)
	if len(parts) == 2 {
		num, err1 := strconv.ParseFloat(strings.TrimSpace(parts[0]), 64)
		den, err2 := strconv.ParseFloat(strings.TrimSpace(parts[1]), 64)
		if err1 == nil && err2 == nil && den != 0 {
			return num / den
		}
	}
	if f, err := strconv.ParseFloat(s, 64); err == nil {
		return f
	}
	return 0
}

func toString(v interface{}) string {
	if v == nil {
		return ""
	}
	return fmt.Sprintf("%v", v)
}

func parseFloat(v interface{}) float64 {
	if v == nil {
		return 0
	}
	switch x := v.(type) {
	case float64:
		return x
	case string:
		f, _ := strconv.ParseFloat(x, 64)
		return f
	}
	return 0
}

func toInt(v interface{}) int {
	if v == nil {
		return 0
	}
	switch x := v.(type) {
	case float64:
		return int(x)
	case string:
		n, _ := strconv.Atoi(x)
		return n
	}
	return 0
}
